### PR TITLE
TST: IntegerNA Support for DataFrame.diff()

### LINKED ISSUE #24171: IntegerNA Support for DataFrame.diff()
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd, numpy as np

df = pd.DataFrame({'a': np.arange(1,6),
                   'b': np.arange(1,6)+2,
                   'c': np.arange(1,6)**2})
print(f"all columns are int: \n{df}")
df.loc[2,'b'] = np.nan
print(f"if one value in column b is set to NA, then the whole column in set to float:\n{df}")
print(f'column b is float, and column a is int, then there is no result for anything for column b-a:\n{df.diff(axis=1)}')
print(f'if the whole dataframe is casted to float, then diff works:\n{df.astype(np.float).diff(axis=1)}')

```
#### Problem description
If for example, in the dataframe of integers, we set (2,b) as NA, then the entire column b becomes float. When we attempt to use df.diff(axis=1), then column b-a gives all NaN because float - int gives NaN. 
In addition, c-b does not even happen, c-a is done instead as only c and a columns are integers. 
Both of these appear unexpected results as in regular python, float - int is a float, and providing results in column c as c-a instead of c-b is conflicting with the definition of the diff function period = 1. 

The workaround to this problem is to cast the entire dataframe to float ```df.astype(np.float).diff(axis=1) ``` and it works. I think this should be fixed in pandas or at least a warning should be given. 

Here is steps by step. Simple dataframe of integers:
```
   a  b   c
0  1  3   1
1  2  4   4
2  3  5   9
3  4  6  16
4  5  7  25
```

If one value in column b is set to NA, then the whole column in set to float:
```
   a    b   c
0  1  3.0   1
1  2  4.0   4
2  3  NaN   9
3  4  6.0  16
4  5  7.0  25
```
when we attempt to do ```df.diff(axis=1)```, then column b-a gives all NaN, however, instead of doing c-b, c-a is done instead, it is only doing difference between the same type columns. For example, in the row 4, 25 - 5 = 20, which means c-a was done instead of c-b:
```
    a   b     c
0 NaN NaN   0.0
1 NaN NaN   2.0
2 NaN NaN   6.0
3 NaN NaN  12.0
4 NaN NaN  20.0
```

if the whole dataframe is casted to float, then diff works as expected:
```
    a    b     c
0 NaN  2.0  -2.0
1 NaN  2.0   0.0
2 NaN  NaN   NaN
3 NaN  2.0  10.0
4 NaN  2.0  18.0
```
panda version 0.23.4. This issue appears in prior versions as well.

### PR DESCRIPTION
**Pull request by Marvzinc**
- [x] closes #24171
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Test for IntegerNA support DataFrame.diff().
``` bash
pytest tests/frame/methods/test_diff.py::TestDataFrameDiff::test_diff_integer_na
```

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the contribution! 
Can you also checked the expected result when doing `diff(axis=1)` ?

File: pandas/tests/frame/methods/test_diff.py [hunk 1]
```diff
@@ -169,3 +169,28 @@ def test_diff_sparse(self):
         )
 
         tm.assert_frame_equal(result, expected)
+
+    def test_diff_integer_na(self):
```
Comment (jreback) [inline pandas/tests/frame/methods/test_diff.py:173]: the OP also was testing for int64, can you parameterize on that as well (result is of course float64 in that case)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_diff.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/tests/frame/methods/test_diff.py:173]: Numpy ints are already tested higher up this file, so I think it is fine to focus here on nullable int

Comment (jreback) [reply pandas/tests/frame/methods/test_diff.py:173]: ok that's fine, let's co-locate the tests then

File: pandas/tests/frame/methods/test_diff.py [hunk 2]
```diff
@@ -169,3 +169,41 @@ def test_diff_sparse(self):
...
+                "a": np.repeat([0, 1, np.nan, 2], 2),
+                "b": np.tile([0, 1, np.nan, 2], 2),
+                "c": np.repeat(np.nan, 8),
+                "d": np.arange(1, 9) ** 2,
+            },
+            dtype="Int64",
+        )
+
+        # Test case for default behaviour of diff
+        result_default = df.diff()
```
Comment (jreback) [inline pandas/tests/frame/methods/test_diff.py:186]: hate to be pendantic about this, but can you use result & expected for both cases (the variable names)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_diff.py (hunk 2 above)
Comment (Marvzinc) [reply pandas/tests/frame/methods/test_diff.py:186]: Sure. Didn't want to overwrite the variable names here. Would you suggest to overwrite, or create a separate test function?

Comment (jorisvandenbossche) [reply pandas/tests/frame/methods/test_diff.py:186]: just overwriting the same names is fine here

Comment (Marvzinc) [reply pandas/tests/frame/methods/test_diff.py:186]: I parameterized the test so overwriting is not necessary anymore.

Comment (MarcoGorelli) [reply pandas/tests/frame/methods/test_diff.py:186]: > I parameterized the test so overwriting is not necessary anymore.

Did you push these changes? Currently I can't see any parametrisation

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks good (will merge shortly, trying to get master to run first).

**Comment by jreback** [comment]
thanks @Marvzinc

**Pull request merged**